### PR TITLE
feat: add a block distance after which we set no TTL in redis

### PR DESCRIPTION
**Pull request by mrice32**
This changes the behavior of the provider caching system s.t. after a certain per-chain block distance in the past, we add no TTL so the cache stores it forever until it is forcibly evicted by redis's max mem eviction policy.

### CODE REVIEW

**Review by nicholaspai — APPROVED** [review]

File: src/utils/ProviderUtils.ts [hunk 1]
```diff
@@ -528,6 +562,12 @@ export async function getProvider(chainId: number, logger?: winston.Logger, useC
 
   const nodeMaxConcurrency = Number(process.env[`NODE_MAX_CONCURRENCY_${chainId}`] || NODE_MAX_CONCURRENCY || "25");
 
+  const disableNoTtl = process.env["DISABLE_NO_TTL"] === "true";
+
+  const noTtlBlockDistance = process.env[`NO_TTL_BLOCK_DISTANCE_${chainId}`]
+    ? Number(process.env[`NO_TTL_BLOCK_DISTANCE_${chainId}`])
+    : DEFAULT_NO_TTL_DISTANCE[chainId];
```
Comment (pxrl) [inline src/utils/ProviderUtils.ts:569]: Minor style nit - IMO preferable to reduce the number of times we have to specify an env var as a string.

```suggestion
  const noTtlBlockDistance = Number(process.env[`NO_TTL_BLOCK_DISTANCE_${chainId}`] ?? DEFAULT_NO_TTL_DISTANCE[chainId]);
```

**Review by pxrl — APPROVED** [review]
Optimistic that this will give a decent performance boost in the bots...

File: src/common/Constants.ts [hunk 2]
```diff
@@ -194,6 +194,20 @@ export const CHAIN_CACHE_FOLLOW_DISTANCE: { [chainId: number]: number } = {
...
+  1: 14400,
+  10: 86400,
+  137: 86400,
+  288: 86400,
+  324: 172800,
+  8453: 86400,
+  42161: 691200,
+  534352: 57600,
+};
+
```
Comment (james-a-morris) [inline src/common/Constants.ts:210]: Do we also want to account for testnets here?

**Review by james-a-morris — APPROVED** [review]
One question. Otherwise LGTM

File: src/utils/ProviderUtils.ts [hunk 3]
```diff
@@ -116,6 +123,8 @@ class CacheProvider extends RateLimitedProvider {
   constructor(
     providerCacheNamespace: string,
     readonly redisClient?: RedisClient,
+    // Note: if not provided, this is set to POSITIVE_INFINITY, meaning the TTL is infinite (i.e. no TTL).
+    readonly noTtlBlockDistance: number = Number.POSITIVE_INFINITY,
```
[the remark this answers is not in the document]
Comment (mrice32) [reply src/utils/ProviderUtils.ts:127]: Probably not. Good call!

File: src/utils/ProviderUtils.ts (hunk 1 above)
Comment (mrice32) [reply src/utils/ProviderUtils.ts]: Actually, I think I want this to be undefined if _neither_ the env or the default is defined. So I don't think we want to wrap both in Number() -- does that make sense? I should add a comment/type annotation about this.

File: src/common/Constants.ts (hunk 2 above)
Comment (mrice32) [reply src/common/Constants.ts:210]: My thought was that for testnets we assume by default we don't want to indefinitely cache anything. If you do want that, you can provide the env. Thoughts?

File: src/utils/ProviderUtils.ts (hunk 1 above)
Comment (mrice32) [reply src/utils/ProviderUtils.ts]: I'm going to go ahead and merge today, but I'll send you a message IRL so you get a chance to come back to this comment.

File: src/common/Constants.ts (hunk 2 above)
Comment (james-a-morris) [reply src/common/Constants.ts:210]: Fair!

File: src/utils/ProviderUtils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrice32) [reply src/utils/ProviderUtils.ts]: Fixed

**Pull request merged**